chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+//
+// The selector can be one of:
+//   - "pending": the chain head
+//   - "latest": the TipSet with the latest executed messages (head - 1)
+//   - "safe": the TipSet with messages executed at least eth.SafeEpochDelay (30) epochs ago or the
+//     latest F3 finalized TipSet
+//   - "finalized": the TipSet with messages executed at least policy.ChainFinality (900) epochs ago
+//     or the latest F3 finalized TipSet
+//   - a decimal block number: the TipSet at the given height
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	switch selector {
+	case EthBlockSelectorEarliest:
+		return nil, fmt.Errorf(`block param "%s" is not supported`, EthBlockSelectorEarliest)
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
+			return tsr.loader.LoadTipSet(ctx, f3TipSet.Parents())
+		} // else F3 is disabled, not running, or behind the default safe or finalized height
+
+		ts, err := tsr.loader.GetTipsetByHeight(ctx, defaultHeight, head, true)
+		if err != nil {
+			return nil, xerrors.Errorf("loading tipset at height %v: %w", defaultHeight, err)
+		}
+		return ts, nil
+
+	default:
+		// likely an 0x hex block number or a decimal block number
+		return tsr.resolveEthBlockNumberSelector(ctx, selector, strict)
+	}
+}
+
+func (tsr *tipSetResolver) getF3FinalizedTipSet(ctx context.Context) (*types.TipSet, error) {
+	if !tsr.f3.IsEnabled() {
+		return nil, nil
+	}
+	if running, _ := tsr.f3.IsRunning(); !running {
+		return nil, nil
+	}
+
+	cert, err := tsr.f3.GetLatestCert(ctx)
+	if err != nil {
+		log.Debugf("loading latest F3 certificate: %s", err)
+		return nil, nil
+	}
+
+	tsk, err := types.TipSetKeyFromBytes(cert.ECChain.Head().Key)
```

Comment:
in the F3 code there's this in `Head()`:

```
	if c.IsZero() {
		return nil
	}
```

I probably should account for this case, what does it mean for this to be the case? Is it a bootstrap thing?

Reply:
We have a gpbft protocol case where the chain can be empty. Finality certs are always guaranteed to have at least one tipset in them.